rp2040-hal/src/pwm/mod.rs

```diff
@@ -0,0 +1,581 @@
+
+mod reg;
+
+use reg::RegisterInterface;
+
+/// Used to pin traits to a specific channel (A or B)
+pub trait ChannelId: Sealed {
+    /// Corresponding [`DynChannelId`](super::DynChannelId)
+    const DYN: DynChannelId;
+}
+
+/// Channel A
+///
+/// These are attached to the even gpio pins and can only do PWM output
+pub enum A {}
+
+/// Channel B
+///
+/// These are attached to the odd gpio pins and can do PWM output and edge counting for input
+pub enum B {}
+
+impl ChannelId for A {
+    const DYN: DynChannelId = DynChannelId::A;
+}
+impl ChannelId for B {
+    const DYN: DynChannelId = DynChannelId::B;
+}
+impl Sealed for A {}
+impl Sealed for B {}
+
+/// Counter is free-running, and will count continuously whenever the slice is enabled
+pub struct FreeRunning;
+/// Count continuously when a high level is detected on the B pin
+pub struct InputHighRunning;
+/// Count once with each rising edge detected on the B pin
+pub struct CountRisingEdge;
+/// Count once with each falling edge detected on the B pin
+pub struct CountFallingEdge;
+
+/// Type-level marker for tracking which slice modes are valid for which slices
+pub trait ValidSliceMode<I: SliceId>: Sealed {}
+
+/// Type-level marker for tracking which slice modes are valid for which slices
+pub trait ValidSliceInputMode<I: SliceId>: Sealed + ValidSliceMode<I> {}
+
+/// Mode for slice
+pub trait SliceMode: Sealed + Sized {
+    /// Corresponding [`DynSliceMode`](super::DynSliceMode)
+    const DYN: DynSliceMode;
+}
+
+impl Sealed for FreeRunning {}
+impl SliceMode for FreeRunning {
+    const DYN: DynSliceMode = DynSliceMode::FreeRunning;
+}
+impl Sealed for InputHighRunning {}
+impl SliceMode for InputHighRunning {
+    const DYN: DynSliceMode = DynSliceMode::InputHighRunning;
+}
+impl Sealed for CountRisingEdge {}
+impl SliceMode for CountRisingEdge {
+    const DYN: DynSliceMode = DynSliceMode::CountRisingEdge;
+}
+impl Sealed for CountFallingEdge {}
+impl SliceMode for CountFallingEdge {
+    const DYN: DynSliceMode = DynSliceMode::CountFallingEdge;
+}
+
+impl<I: SliceId> ValidSliceMode<I> for FreeRunning {}
+impl<I: SliceId> ValidSliceMode<I> for InputHighRunning {}
+impl<I: SliceId> ValidSliceMode<I> for CountRisingEdge {}
+impl<I: SliceId> ValidSliceMode<I> for CountFallingEdge {}
+impl<I: SliceId> ValidSliceInputMode<I> for InputHighRunning {}
+impl<I: SliceId> ValidSliceInputMode<I> for CountRisingEdge {}
+impl<I: SliceId> ValidSliceInputMode<I> for CountFallingEdge {}
+
+//==============================================================================
+//  Slice IDs
+//==============================================================================
+
+/// Type-level `enum` for slice IDs
+pub trait SliceId: Sealed {
+    /// Corresponding [`DynSliceId`](super::DynSliceId)
+    const DYN: DynSliceId;
+    /// [`DivMode`] at reset
+    type Reset;
+}
+
+macro_rules! slice_id {
+    ($Id:ident, $NUM:literal, $reset : ident) => {
+        #[doc = "Slice ID representing slice "]
```

Comment:
The documentation could contain the numeric ID.
```suggestion
        #[doc = std::concat!("Slice ID representing slice ", $NUM)]
```